src/modules/fw_att_control/FixedwingAttitudeControl.cpp

```diff
@@ -638,6 +656,48 @@ void FixedwingAttitudeControl::run()
 				control_input.groundspeed = groundspeed;
 				control_input.groundspeed_scaler = groundspeed_scaler;
 
+				/* bi-linear interpolation over airspeed for actuator trims */
+				float trim_roll = _parameters.trim_roll;
+				float trim_pitch = _parameters.trim_pitch;
+				float trim_yaw = _parameters.trim_yaw;
+
+				if (airspeed < _parameters.airspeed_trim) {
+					if (fabsf(_parameters.trim_roll_vmin) >= 0.01f || fabsf(_parameters.trim_roll_vmax) >= 0.01f) {
```

Comment:
Isn't 0 safe here?

Reply:
the idea with this check is that the trims should be monotonically increasing or decreasing. so if someone leaves them both zero - it effectively just means that whatever the nominal trim was set to is the only trim for all airspeeds (the standard case up until now). this is mostly just to not effect anyone's current setup if they have trims they like that arent scheduled, and maybe don't notice the new parameters. if either of the new trims (vmin or vmax) are set to another value, we are assuming the user specifically did this because they wanted to - and then it is their responsibility to make sure the values make sense from that point.

Reply:
to maybe explain better - think of the case that someone doesnt know these trims exist.. and they have a nominal trim of 0.15 on the elevator, if we dont have this check to disable the new trims -- then the controller would be reducing the elevator trim to 0 in both minimum and maximum airspeed cases -- which i have a hard time thinking of a plane that exhibits this non monotonic trimming behavior

Reply:
I'm simply suggesting skipping the checks and always calling the gradual. The result should be the same, and it's easier to read, less code, and negligible computationally.

Reply:
If we skip the checks, then the unset trim vmin and vmax will be used in the gradual. That is what I'm trying to avoid. The checks are necessary regardless of if we skip the gradual, or, when the checks show trim vmax and vmin disabled, redefining the vmax and vmin trims as equal to the nominal trim and proceeding through the gradual with a flat line to interpolate on.

Reply:
@bresch I actually originally had it this way, I also prefer to put it there. @dagar wanted it moved to where the trims were applied. I'm a bit indifferent in the end as long as it works (either one does the same function) -- can you two decide which you like?

Reply:
Sorry, I didn't mean to make this so complicated. The problem I have with this kind of thing is giving 0 some special meaning when I can think of cases (contrived?) where using 0 as 0 would be valid.

Isn't it conceivable someone would want 0 pitch trim at max airspeed? How are they going to know where the line is between a tiny value near 0 and 0 as interpreted by the controller that actually results in using TRIM_PITCH?

Reply:
@dagar I understand your concern if this was only one parameter in the condition. The point here is that it takes **both max and min params to be set to 0 for them not to be used. the only case where this would be an issue, is if someone actually wanted min=0, max=0, and nom=anything else. But I would argue (as i did previously) that this will *never be the case for anyone properly using airspeed dependent trims. they should be, in the nominal operating region, monotonic - which means the 0 - some value - 0 case will not happen. otherwise -- if someone who has one trim value set (only the nominal) doesnt want to even bother with airspeed dependent trims, leaving the min and max at the default of zero means it wont affect him/her.

this does *not prohibit someone from using trim_pitch_max = 0, so long as trim_pitch_min is set to anything else that makes sense (i.e., anything besides 0 -- if trim_pitch_min=0 as well.. then trim_pitch_nom should also=0 -- again the monotonic argument).

Reply:
> if (_parameters.trim_roll_vmin < 0.01f) {
>    _parameters.trim_roll_vmin = _parameters.trim_roll;
> }

also -- just to be clear / make sure we are on the same page -- @bresch forgot the second condition in this if statement in his comment -- i.e. I originally had it like this: 

```
if ( fabsf(_parameters.trim_roll_vmin) < 0.01f && fabsf(_parameters.trim_roll_max) < 0.01f ) {
    _parameters.trim_roll_vmin = _parameters.trim_roll;
    _parameters.trim_roll_vmax = _parameters.trim_roll;
}
```

If put up where the parameters are defined.

Reply:
Yes that works, it's just unintuitive. So you set the max to 0 and it's working as expected, later you decide to kill the min trim because you didn't want it for whatever reason, but now your max has also gone away? I don't mean to be so difficult here, I'm trying to anticipate all the various ways things are misused, and won't' necessarily make sense until reading the code.

What if we applied these relative to the main TRIM? TRIM_X is your baseline, then FW_TRIM_X_VMIN/FW_TRIM_X_VMAX are some amount of additional trim applied linearly as you approach the min/max airspeeds.

    roll min = TRIM_ROLL + FW_TRIM_R_VMIN
    roll trim = TRIM_ROLL
    roll max = TRIM_ROLL + FW_TRIM_R_VMAX

Reply:
sure. this can also work. maybe also more robust to people not understanding exactly what to expect. then we get rid of the checks all together, and just assume that if someone sets the min/max trim deltas different from zero -- they should know what they are doing.

if these are now added deltas to the nominal trims, like the ones for flaps -- should we then change names to "DTRIM" (or anything similar)?

Reply:
Great, yes whichever name you think makes the most sense. DTRIM or something mentioning additional trim.